### PR TITLE
Login Changes

### PR DESCRIPTION
**Pull request by xenni**
### non-premium
- warns non-premium users that their accounts are not premium and fails to launch
### offline authentication
- uses lastlogin file to allow offline users to authenticate if they have authenticated previously
- uses SHA256 to ensure that the user's password cannot be recovered from the file, unless they checked 'Remember', in which case the password is stored as usual.
### other
- adds '.idea' to '.gitignore' to ignore IntelliJ IDEA project files

### CODE REVIEW

File: src/main/java/org/spoutcraft/launcher/gui/LoginForm.java [hunk 1]
```diff
@@ -602,60 +624,59 @@ public class LoginForm extends JFrame implements ActionListener, DownloadListene
 			protected void done() {
 				if (values == null || values.length < 4)
 					return;
-				gu.user = values[2].trim();
```
Comment (xenni) [inline src/main/java/org/spoutcraft/launcher/gui/LoginForm.java:605]: Hehe. Move one line and you generate ~50 deletions and insertions. I love diff.

File: src/main/java/org/spoutcraft/launcher/gui/LoginForm.java [hunk 2]
```diff
@@ -74,14 +74,12 @@ import javax.swing.SwingConstants;
 import javax.swing.SwingWorker;
 import javax.swing.border.EmptyBorder;
 
-import org.spoutcraft.launcher.GameUpdater;
-import org.spoutcraft.launcher.MinecraftUtils;
-import org.spoutcraft.launcher.PlatformUtils;
-import org.spoutcraft.launcher.SettingsHandler;
+import org.spoutcraft.launcher.*;
```
Comment (xenni) [inline src/main/java/org/spoutcraft/launcher/gui/LoginForm.java:77]: automatically collapsed by IntelliJ

**Comment by Top-Cat** [comment]
You realise there is already a lastlogin file that stores multiple logins, right?

**Comment by xenni** [comment]
Really? I noticed the file being generated, but I figured Minecraft was the one doing it.
I did also poke around in the code fairly extensively, is it named something unintuitive?
Where is it located?

**Comment by Top-Cat** [comment]
**Read:**
https://github.com/SpoutDev/Spoutcraft-Launcher/blob/master/src/main/java/org/spoutcraft/launcher/gui/LoginForm.java#L458

---

**Write:**
https://github.com/SpoutDev/Spoutcraft-Launcher/blob/master/src/main/java/org/spoutcraft/launcher/gui/LoginForm.java#L506

---

I don't mind the encryption changing, just having two files seems silly and everyone's data is already here

**Comment by xenni** [comment]
So that means that those entries are only written if login was successful, correct?
If so, then I'll fix my code immediately to use that.

**Comment by Top-Cat** [comment]
Exactly, except it doesn't save their password unless they checked the box which is correct behaviour.

**Comment by xenni** [comment]
Okay. That commit now uses the builtin lastlogin file for storing (encrypted) passwords if 'Remember' is checked or the password's SHA256 hash, if 'Remember' is not checked. I've tested it and it appears to work quite well.

**Comment by xenni** [comment]
Oh, this does change the format of the lastlogin file in a breaking way. How do you think it would be best to handle old versions of the file?

File: src/main/java/org/spoutcraft/launcher/gui/LoginForm.java [hunk 3]
```diff
@@ -474,22 +490,31 @@ public class LoginForm extends JFrame implements ActionListener, DownloadListene
...
-							drawCharacter("http://s3.amazonaws.com/MinecraftSkins/" + user + ".png", 103, 170, loginSkin1Image);
-						} else if (i == 2) {
-							loginSkin2.setText(user);
-							loginSkin2.setVisible(true);
-							drawCharacter("http://s3.amazonaws.com/MinecraftSkins/" + user + ".png", 293, 170, loginSkin2Image);
-						}
-					}
-
-					usernames.put(user, pass);
+                    boolean isHash = dis.readBoolean();
```
Comment (xenni) [inline src/main/java/org/spoutcraft/launcher/gui/LoginForm.java:490]: This line breaks compatibility with the older version of the lastlogin file. It is nessecary, however, to differentiate between text passwords and SHA256 digests.

**Comment by Top-Cat** [comment]
You seem to be reinventing the wheel, the lastlogins file already stores authed users and the SHA256 seems like it just breaks compatibility...

**Comment by xenni** [comment]
It's simple to change, but I figured it would be best if we did not store passwords unless the user actually wanted them stored.
It does break compatibility, but perhaps we could just name the new version of the file something different.
Users would have to re-enter credentials, but we could also just read the old version in if it exists, and delete it when the new version has been created.

**Comment by Top-Cat** [comment]
We don't store passwords if they don't want them stored. The username is always tho.

**Comment by xenni** [comment]
Right. This is how it operates currently. The only difference is that we now store the SHA256 hash of the password if 'Remember' was not selected. That way, we can check against it when they try to authenticate offline.

Is there anything wrong with the check-for-old-file, create-new-file-with-different-name scheme?

**Comment by Top-Cat** [comment]
Storing a hash when it's not requested isn't a good idea, and what are you checking it against?

**Comment by xenni** [comment]
I think we are misunderstanding eachother.

If we store the hash, it cannot be undone to retrieve the pasword, so it isn't 'remembered'.

If the user checks 'Remember', then we store the encrypted password as usual.
However, if they don't check remember then we store the SHA256 hash, that way we can ensure they are using the same password the next time they try to login and we are unable to verify the password against login.minecraft.net.

The SHA256 hashing is nessecary to handle the following usage case:
 1) user logs in, and the account is verified against login.minecraft.net ('Remember' is not selected)
 2) user plays, etc... and eventually exists minecraft
 3) user looses internet connection or minecraft.net goes down
 4) user wants to play minecraft
 5) user enters username and password and attempts to login again

Because we cannot verify the password that the user enters againt minecraft.net, we need some way to ensure the proper password was entered, otherwise anyone could just enter anything into the password box.

We use SHA256 to store the digest of the password, that way it cannot be recovered from the lastlogin file.

**Comment by lukespragg** [comment]
Why were the imports changed to a wildcard?

**Comment by xenni** [comment]
It was automatically done by IntelliJ.

**Comment by Afforess** [comment]
@xenni - okay, I understand your rational. I would appreciate it if you could clean up this pull up a bit, you change brace styles a few times. And once you clean it up, please rebase it into 1 commit. I will pull it then.

**Comment by xenni** [comment]
I did the syntax cleanup, but i'm at a loss at how to do the rebase. I've never really worked much with Git before, to be honest. The docs on 'rebase' only served to make me even more confused.

**Comment by Afforess** [comment]
Merged manually.

**Pull request merged**